

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>1. Analysis workflow &mdash; Spiker documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="1. Prepare FASTQ files" href="walkthrough.html" />
    <link rel="prev" title="1. Prerequisites" href="installation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: white" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> Spiker
          

          
            
            <img src="_static/logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Introduction</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction to spike-in control</a></li>
</ul>
<p class="caption"><span class="caption-text">Installation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">1. Prerequisites</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html#python-dependencies">2. Python Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html#install-spiker">3. Install Spiker</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html#upgrade-spiker">4. Upgrade Spiker</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html#uninstall">5. Uninstall</a></li>
</ul>
<p class="caption"><span class="caption-text">Usage information</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">1. Analysis workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="#input-and-output">2. Input and output</a></li>
<li class="toctree-l1"><a class="reference internal" href="#spiker-py-options">3. Spiker.py options</a></li>
<li class="toctree-l1"><a class="reference internal" href="#split-bam-py-options">4. split_bam.py options</a></li>
</ul>
<p class="caption"><span class="caption-text">Walkthrough example</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="walkthrough.html">1. Prepare FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="walkthrough.html#build-bowtie2-index-files">2. Build bowtie2 index files</a></li>
<li class="toctree-l1"><a class="reference internal" href="walkthrough.html#map-reads-to-the-composite-reference-genome">3. Map reads to the composite reference genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="walkthrough.html#calculate-scaling-factors">4. Calculate scaling factors</a></li>
<li class="toctree-l1"><a class="reference internal" href="walkthrough.html#run-spiker">5. Run Spiker</a></li>
</ul>
<p class="caption"><span class="caption-text">Bowtie2 Index Files</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="bowtie2_index.html">1. Bowtie2 index files</a></li>
<li class="toctree-l1"><a class="reference internal" href="bowtie2_index.html#index-files-for-human-drosophila">2. Index files for Human + Drosophila</a></li>
<li class="toctree-l1"><a class="reference internal" href="bowtie2_index.html#index-files-for-mouse-drosophila">3. Index files for Mouse + Drosophila</a></li>
</ul>
<p class="caption"><span class="caption-text">Test Dataset</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="dataset.html">1. Test dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="dataset.html#h3k79me2-chip-seq-data">2. H3K79me2 ChIP-seq data</a></li>
<li class="toctree-l1"><a class="reference internal" href="dataset.html#wce-whole-cell-extract-data">3. WCE (whole cell extract) data</a></li>
<li class="toctree-l1"><a class="reference internal" href="dataset.html#md5sum">4. MD5sum</a></li>
</ul>
<p class="caption"><span class="caption-text">Release history</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="change_log.html">1. Release history</a></li>
</ul>
<p class="caption"><span class="caption-text">Liscense</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="liscense.html">LICENSE</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Spiker</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li><span class="section-number">1. </span>Analysis workflow</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/usage.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="analysis-workflow">
<h1><span class="section-number">1. </span>Analysis workflow<a class="headerlink" href="#analysis-workflow" title="Permalink to this headline">¶</a></h1>
<p><strong>Spiker</strong> is specifically designed to analyze ChIP-seq data with spike-in control.</p>
<a class="reference internal image-reference" href="_images/workflow.png"><img alt="Alternative text" src="_images/workflow.png" style="width: 600px;" /></a>
</div>
<div class="section" id="input-and-output">
<h1><span class="section-number">2. </span>Input and output<a class="headerlink" href="#input-and-output" title="Permalink to this headline">¶</a></h1>
<p>The input files can be single-end or paired-end <a class="reference external" href="https://en.wikipedia.org/wiki/FASTQ_format#">FASTQ</a> files or pre-aligned <a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/bam.html">BAM</a> files. It supports both narrow peak (such as H3K27ac) and broad peak analysis (such as H3K79me2 in this tutorial).</p>
<p>Input <strong>single-end</strong> fastq files</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ spiker.py --t1 H3K27ac.fastq.gz --c1 control.fastq.gz --bt2-index /data/GRCh38 --spikeIn --csf 1.23 --tsf 0.95 -o H3K27ac
$ spiker.py --broad --t1 H3K27ac.fastq.gz --c1 control.fastq.gz --bt2-index /data/GRCh38 --spikeIn --csf 1.23 --tsf 0.95 -o H3K27ac
</pre></div>
</div>
<p>Input <strong>paired-end</strong> fastq files</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ spiker.py --t1 H3K27ac_R1.fastq.gz --t2 H3K27ac_R2.fastq.gz --c1 control_R1.fastq.gz --c2 control_R2.fastq.gz --bt2-index /data/GRCh38 --spikeIn --csf 1.23 --tsf 0.95 -o H3K27ac
$ spiker.py --broad --t1 H3K27ac_R1.fastq.gz --t2 H3K27ac_R2.fastq.gz --c1 control_R1.fastq.gz --c2 control_R2.fastq.gz --bt2-index /data/GRCh38 --spikeIn --csf 1.23 --tsf 0.95 -o H3K27ac
</pre></div>
</div>
<p>Input <strong>BAM</strong> files</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ spiker.py -t H3K27ac.sorted.bam -c control.sorted.bam --spikeIn --csf 1.23 --tsf 0.95 -o H3K27ac
$ spiker.py --broad -t H3K27ac.sorted.bam -c control.sorted.bam --spikeIn --csf 1.23 --tsf 0.95 -o H3K27ac
</pre></div>
</div>
<p><strong>Output files</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format12">ENCODE narrowPeak</a></p></li>
<li><p><a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format14">ENCODE gappedPeak</a></p></li>
<li><p><a class="reference external" href="https://genome.ucsc.edu/goldenpath/help/bigWig.html">bigWig</a></p></li>
<li><p><a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/bedgraph.html">bedGraph</a></p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="spiker-py-options">
<h1><span class="section-number">3. </span>Spiker.py options<a class="headerlink" href="#spiker-py-options" title="Permalink to this headline">¶</a></h1>
<dl>
<dt>Options:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">--version</span></kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
<dt><kbd><span class="option">--t1=<var>CHIP_R1</var></span></kbd></dt>
<dd><p>FASTQ file (read1) for ChIP sample. Can be regular
plain text file or compressed file (.gz, .bz2).
Mutually exclusive with ‘-t’.</p>
</dd>
<dt><kbd><span class="option">--t2=<var>CHIP_R2</var></span></kbd></dt>
<dd><p>FASTQ file (reas2) for ChIP sample. Can be regular
plain text file or compressed file (.gz, .bz2).
Mutually exclusive with ‘-t’. Ignore this for single-
end sequencing.</p>
</dd>
<dt><kbd><span class="option">-t <var>CHIP_BAM</var></span>, <span class="option">--treat=<var>CHIP_BAM</var></span></kbd></dt>
<dd><p>BAM file of ChIP sample. The BAM file must be sorted
and indexed. Mutually exclusive with ‘–t1’ and ‘–
t2’.</p>
</dd>
<dt><kbd><span class="option">--c1=<var>CTRL_R1</var></span></kbd></dt>
<dd><p>FASTQ file (read1) for Control sample. Can be regular
plain text file or compressed file (.gz, .bz2).
Mutually exclusive with ‘-c’.</p>
</dd>
<dt><kbd><span class="option">--c2=<var>CTRL_R2</var></span></kbd></dt>
<dd><p>FASTQ file (reas2) for Control sample. Can be regular
plain text file or compressed file (.gz, .bz2).
Mutually exclusive with ‘-c’. Ignore this for single-
end sequencing.</p>
</dd>
<dt><kbd><span class="option">-c <var>CTRL_BAM</var></span>, <span class="option">--control=<var>CTRL_BAM</var></span></kbd></dt>
<dd><p>BAM file of Control sample. Mutually exclusive with ‘
–c1’ and ‘–c2’. The BAM file must be sorted and
indexed.</p>
</dd>
<dt><kbd><span class="option">-o <var>OUTFILE</var></span>, <span class="option">--output=<var>OUTFILE</var></span></kbd></dt>
<dd><p>Prefix of output files.</p>
</dd>
<dt><kbd><span class="option">--bt2-index=<var>BT2_INDEX</var></span></kbd></dt>
<dd><p>The prefix (minus trailing .X.bt2) for bowtie2 index
files. Ignore this option if BAM files were provided
by ‘-t’ and ‘-c’.</p>
</dd>
<dt><kbd><span class="option">-n <var>N_READS</var></span></kbd></dt>
<dd><p>Number of alignments from the BAM file used to tell
the sequencing layout (PE or SE), and estiamte the
fragment size ‘d’. default=1000000</p>
</dd>
<dt><kbd><span class="option">-g <var>G_SIZE</var></span>, <span class="option">--genome-size=<var>G_SIZE</var></span></kbd></dt>
<dd><p>Effective genome size. It can be 1.0e+9 or 1000000000,
or shortcuts:’hs’ for human (2.7e9), ‘mm’ for mouse
(1.87e9), ‘ce’ for C. elegans (9e7) and ‘dm’ for
fruitfly (1.2e8). default=hs</p>
</dd>
<dt><kbd><span class="option">-p <var>N_THREADS</var></span>, <span class="option">--proc=<var>N_THREADS</var></span></kbd></dt>
<dd><p>Number of threads. default=8</p>
</dd>
<dt><kbd><span class="option">--mfold=<var>M_FOLD</var></span></kbd></dt>
<dd><p>Select the regions within MFOLD range of high-
confidence enrichment ratio against background to
build model. Fold-enrichment in regions must be lower
than upper limit, and higher than the lower limit. Use
as “-m 10 30”. DEFAULT:5 50</p>
</dd>
<dt><kbd><span class="option">--spikeIn</span></kbd></dt>
<dd><p>Set this flag if ChIP and control samples contains
exogenous reads as splike-in. Please note, you also
need to specify –tsf and –csf.</p>
</dd>
<dt><kbd><span class="option">--tsf=<var>TREAT_SF</var></span></kbd></dt>
<dd><p>Scaling factor for treatment. This will be applied to
the pileup bedgraph file of treatment
(<a href="#id1"><span class="problematic" id="id2">*</span></a>.treat.pileup.bdg).</p>
</dd>
<dt><kbd><span class="option">--csf=<var>CONTROL_SF</var></span></kbd></dt>
<dd><p>Scaling factor for control. This will be applied to
the pileup bedgraph file of maximum background
(<a href="#id3"><span class="problematic" id="id4">*</span></a>.control.pileup.max.bdg).</p>
</dd>
<dt><kbd><span class="option">--q-peak=<var>Q_CUTOFF</var></span></kbd></dt>
<dd><p>Qvalue cutoff for peaks. default=0.05</p>
</dd>
<dt><kbd><span class="option">--q-link=<var>Q_LINK_CUT</var></span></kbd></dt>
<dd><p>Qvalue cutoff for linking regions. default=0.1</p>
</dd>
<dt><kbd><span class="option">--bw</span></kbd></dt>
<dd><p>If set, generate bigwig files for ChIP pileup and
control pileup.</p>
</dd>
<dt><kbd><span class="option">--maxgap=<var>MAX_GAP</var></span></kbd></dt>
<dd><p>maximum gap between significant points in a peak.
default=100</p>
</dd>
<dt><kbd><span class="option">--broad</span></kbd></dt>
<dd><p>If set, call broad peaks.</p>
</dd>
<dt><kbd><span class="option">--frip</span></kbd></dt>
<dd><p>If set, calculate FRiP (the Fraction of Reads In
called Peaks) score using the BAM and peak files.</p>
</dd>
<dt><kbd><span class="option">--cleanup</span></kbd></dt>
<dd><p>If set, clean up the intermediate files. When not set,
intermediate files are kept so that rerun the workflwo
will be much faster.</p>
</dd>
<dt><kbd><span class="option">--refine</span></kbd></dt>
<dd><p>If set, detect peak summit position.</p>
</dd>
<dt><kbd><span class="option">--verbose</span></kbd></dt>
<dd><p>If set, print detailed information for debugging.</p>
</dd>
</dl>
</dd>
</dl>
</div>
<div class="section" id="split-bam-py-options">
<h1><span class="section-number">4. </span>split_bam.py options<a class="headerlink" href="#split-bam-py-options" title="Permalink to this headline">¶</a></h1>
<dl>
<dt>Options:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">--version</span></kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
<dt><kbd><span class="option">-i <var>BAM_FILE</var></span></kbd></dt>
<dd><p>BAM file of the composite genome (such as human + fly)</p>
</dd>
<dt><kbd><span class="option">-o <var>OUT_PREFIX</var></span>, <span class="option">--output=<var>OUT_PREFIX</var></span></kbd></dt>
<dd><p>Output prefix. The original BAM file will be split
into four BAM files: ‘prefix_human.bam’,
‘prefix_exogenous.bam’, ‘prefix_both.bam’,
‘prefix_neither.bam’.</p>
</dd>
<dt><kbd><span class="option">-p <var>CHR_PREFIX</var></span>, <span class="option">--exo-prefix=<var>CHR_PREFIX</var></span></kbd></dt>
<dd><p>Prefix added to the exogenous chromosome IDs. For
example. ‘chr2L’ -&gt; ‘dm6_chr2L’. default=dm6_</p>
</dd>
<dt><kbd><span class="option">-q <var>MAP_QUAL</var></span>, <span class="option">--mapq=<var>MAP_QUAL</var></span></kbd></dt>
<dd><p>Mapping quality (phred scaled) threshold. Alignments
with mapping quality score lower than this will be
assigned to ‘prefix_neither.bam’. default=30</p>
</dd>
<dt><kbd><span class="option">--threads=<var>N_THREAD</var></span></kbd></dt>
<dd><p>Number of threads to use for BAM sorting. default=1</p>
</dd>
</dl>
</dd>
</dl>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="walkthrough.html" class="btn btn-neutral float-right" title="1. Prepare FASTQ files" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral float-left" title="1. Prerequisites" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2021, Liguo Wang
      <span class="lastupdated">
        Last updated on Feb 12, 2021.
      </span>

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
    <!-- Theme Analytics -->
    <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-XXXXXXX-1', 'auto');
    ga('send', 'pageview');
    </script>

    
   

</body>
</html>