

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>1. Prepare FASTQ files &mdash; Spiker documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="1. Bowtie2 index files" href="bowtie2_index.html" />
    <link rel="prev" title="1. Spiker" href="usage.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: white" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> Spiker
          

          
            
            <img src="_static/logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Introduction</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
</ul>
<p class="caption"><span class="caption-text">Installation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">1. Prerequisites</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html#python-dependencies">2. Python Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html#install-spiker">3. Install Spiker</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html#upgrade-spiker">4. Upgrade Spiker</a></li>
<li class="toctree-l1"><a class="reference internal" href="change_log.html">5. Release history</a></li>
</ul>
<p class="caption"><span class="caption-text">Usage information</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="usage.html">1. Spiker</a></li>
</ul>
<p class="caption"><span class="caption-text">Walkthrough example</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">1. Prepare FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="#build-bowtie2-index-files">2. Build bowtie2 index files</a></li>
<li class="toctree-l1"><a class="reference internal" href="#map-reads-to-the-composite-reference-genome">3. Map reads to the composite reference genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="#calculate-scaling-factors">4. Calculate scaling factors</a></li>
<li class="toctree-l1"><a class="reference internal" href="#run-spiker">5. Run Spiker</a></li>
</ul>
<p class="caption"><span class="caption-text">Bowtie2 Index Files</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="bowtie2_index.html">1. Bowtie2 index files</a></li>
<li class="toctree-l1"><a class="reference internal" href="bowtie2_index.html#index-files-for-human-drosophila">2. Index files for Human + Drosophila</a></li>
<li class="toctree-l1"><a class="reference internal" href="bowtie2_index.html#index-files-for-mouse-drosophila">3. Index files for Mouse + Drosophila</a></li>
</ul>
<p class="caption"><span class="caption-text">Test Dataset</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="dataset.html">1. Test dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="dataset.html#h3k79me2-chip-seq-data">2. H3K79me2 ChIP-seq data</a></li>
<li class="toctree-l1"><a class="reference internal" href="dataset.html#wce-whole-cell-extract-data">3. WCE (whole cell extract) data</a></li>
<li class="toctree-l1"><a class="reference internal" href="dataset.html#md5sum">4. MD5sum</a></li>
</ul>
<p class="caption"><span class="caption-text">Liscense</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="liscense.html">LICENSE</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Spiker</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li><span class="section-number">1. </span>Prepare FASTQ files</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/walkthrough.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="prepare-fastq-files">
<h1><span class="section-number">1. </span>Prepare FASTQ files<a class="headerlink" href="#prepare-fastq-files" title="Permalink to this headline">¶</a></h1>
<p>Usually, the FASTQ files are delivered to you from the sequencing core. In this tutorial, we will use the H3K79me2 ChIP-seq data published by <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/25437568/">Orlando et al (Cell Reports, 2014)</a>. In this study, Drosophila S2 cells were added to human Jurkat cells at the ratio of 1:2.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 39%" />
<col style="width: 10%" />
<col style="width: 40%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Sample (ChIP)</p></td>
<td><p>SRR_accession</p></td>
<td><p>Sample (WCE)</p></td>
<td><p>SRR_accession</p></td>
</tr>
<tr class="row-even"><td><p>Jurkat_K79_0%_R1</p></td>
<td><p><a class="reference external" href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR1536557">SRR1536557</a></p></td>
<td><p>Jurkat_WCE_0%_R1</p></td>
<td><p><a class="reference external" href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR1584489">SRR1584489</a></p></td>
</tr>
<tr class="row-odd"><td><p>Jurkat_K79_25%_R1</p></td>
<td><p><a class="reference external" href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR1536558">SRR1536558</a></p></td>
<td><p>Jurkat_WCE_25%_R1</p></td>
<td><p><a class="reference external" href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR1584490">SRR1584490</a></p></td>
</tr>
<tr class="row-even"><td><p>Jurkat_K79_50%_R1</p></td>
<td><p><a class="reference external" href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR1536559">SRR1536559</a></p></td>
<td><p>Jurkat_WCE_50%_R1</p></td>
<td><p><a class="reference external" href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR1584491">SRR1584491</a></p></td>
</tr>
<tr class="row-odd"><td><p>Jurkat_K79_75%_R1</p></td>
<td><p><a class="reference external" href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR1536560">SRR1536560</a></p></td>
<td><p>Jurkat_WCE_75%_R1</p></td>
<td><p><a class="reference external" href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR1584492">SRR1584492</a></p></td>
</tr>
<tr class="row-even"><td><p>Jurkat_K79_100%_R1</p></td>
<td><p><a class="reference external" href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR1536561">SRR1536561</a></p></td>
<td><p>Jurkat_WCE_100%_R1</p></td>
<td><p><a class="reference external" href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR1584493">SRR1584493</a></p></td>
</tr>
</tbody>
</table>
<p>After download the SRA files to your hard drive, you will need the <code class="code docutils literal notranslate"><span class="pre">fastq-dump</span></code> command from the <a class="reference external" href="https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc">SRA Tookit</a> to convert the SRA archive files into FASTQ format. We used this command to convert single-end sequencing data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ fastq-dump SRA_file1 SRA_file2 SRA_file3
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Alternatively, you can download FASTQ files <strong>directly</strong> from the <a class="reference external" href="https://www.ebi.ac.uk/ena/browser/home">ENA (Enropean Nucleotide Archive)</a> by searching SRR accession numbers listed in the above table.</p>
</div>
</div>
<div class="section" id="build-bowtie2-index-files">
<h1><span class="section-number">2. </span>Build bowtie2 index files<a class="headerlink" href="#build-bowtie2-index-files" title="Permalink to this headline">¶</a></h1>
<p>To calculate how many reads in the FASTQ files are drived from the Drosophila S2 cells, we could map all reads to the <strong>composite reference genome</strong> (i.e., human + Drosophila). In this tutorial, we will use <a class="reference external" href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">bowtie2</a>, other short reads aligners such as <a class="reference external" href="http://bio-bwa.sourceforge.net/">BWA</a> also work fine. To use <a class="reference external" href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">bowtie2</a> to map reads to the
<strong>composite reference genome</strong>, we need to create bowtie2 index files first.</p>
<p>Download the human and Drosophila reference genome sequences:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
$ wget http://hgdownload.soe.ucsc.edu/goldenPath/dm6/bigZips/dm6.fa.gz
$ gunzip hg38.fa.gz
$ gunzip dm6.fa.gz
</pre></div>
</div>
<p>Rename chromosome IDs of the Drosophila genome:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sed &#39;s/chr/dm6_chr/g&#39; dm6.fa  &gt; dm6_renameID.fa
</pre></div>
</div>
<p>Concatenate the human and Drosophila genome files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat hg38.fa  dm6_renameID.fa &gt; hg38_dm6.fa
</pre></div>
</div>
<p>Run <code class="code docutils literal notranslate"><span class="pre">bowtie2-build</span></code> to build index files. “hg38_dm6” is the prefix of index files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ bowtie2-build  hg38_dm6.fa  hg38_dm6
</pre></div>
</div>
<p>Clean up:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ rm hg38.fa dm6.fa dm6_renameID.fa
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>We have <a class="reference internal" href="bowtie2_index.html#bowtie2-index"><span class="std std-ref">Bowtie2 index files</span></a> files for <strong>hg38 + dm6</strong>, <strong>hg19 + dm6</strong>, <strong>mm9 + dm6</strong>, <strong>mm10 + dm6</strong>, <strong>mm39 + dm6</strong> available for download.</p>
</div>
</div>
<div class="section" id="map-reads-to-the-composite-reference-genome">
<h1><span class="section-number">3. </span>Map reads to the composite reference genome<a class="headerlink" href="#map-reads-to-the-composite-reference-genome" title="Permalink to this headline">¶</a></h1>
<p>Use <a class="reference external" href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">bowtie2</a> to map all reads to the <strong>composite reference genome</strong> (i.e., hg38_dm6.fa). Sort and index the BAM files using <a class="reference external" href="http://www.htslib.org/">Samtools</a>. Below is an example to process one sample:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bowtie2</span> <span class="o">--</span><span class="n">threads</span> <span class="mi">4</span> <span class="o">-</span><span class="n">x</span> <span class="o">./</span><span class="n">database</span><span class="o">/</span><span class="n">GRCh38_dm6</span>  <span class="o">-</span><span class="n">U</span> <span class="n">Jurkat_K79_00p_Rep1_SRR1536557</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="o">|</span> <span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">Sbh</span> <span class="o">-</span> <span class="o">&gt;</span> <span class="n">Jurkat_K79_00p_Rep1_SRR1536557</span><span class="o">.</span><span class="n">bam</span>
<span class="n">samtools</span> <span class="n">sort</span> <span class="o">-@</span> <span class="mi">4</span> <span class="n">Jurkat_K79_00p_Rep1_SRR1536557</span><span class="o">.</span><span class="n">bam</span> <span class="o">&gt;</span> <span class="n">Jurkat_K79_00p_Rep1_SRR1536557</span><span class="o">.</span><span class="n">sorted</span><span class="o">.</span><span class="n">bam</span>
<span class="n">samtools</span> <span class="n">index</span> <span class="o">-@</span> <span class="mi">4</span> <span class="n">Jurkat_K79_00p_Rep1_SRR1536557</span><span class="o">.</span><span class="n">sorted</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>We have pre-aligned BAM files available from <a class="reference internal" href="dataset.html#dataset"><span class="std std-ref">Test dataset</span></a> for download.</p>
</div>
</div>
<div class="section" id="calculate-scaling-factors">
<h1><span class="section-number">4. </span>Calculate scaling factors<a class="headerlink" href="#calculate-scaling-factors" title="Permalink to this headline">¶</a></h1>
<p>To remove experimental bias and accurately quantify ChIP signal changes, the amount of exogenous Drosophila chromatin in different samples needs to be normalized to the same level. After sequencing, this is equivalent to normalize Drosophila reads to the same amount.</p>
<p>We provide <code class="code docutils literal notranslate"><span class="pre">split_bam.py</span></code> to split the BAM file into 4 sub-BAM files.</p>
<blockquote>
<div><ul class="simple">
<li><p>_both.bam : contains reads mapped to both human and Drosophila genomes</p></li>
<li><p>_exogenous.bam : contains reads only map to the Drosophila genomes</p></li>
<li><p>_human.bam : contains reads only map to the human genomes</p></li>
<li><p>_neither.bam : contains qcfailed, low quality, unmapped, dupilcate reads etc.</p></li>
</ul>
</div></blockquote>
<p>A report file describing the number of alignments in each sub-BAM is also generated.</p>
<p>For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ split_bam.py --threads 8  -i Jurkat_K79_00p_Rep1_SRR1536557.sorted.bam  -o Jurkat_K79_00p_Rep1
Read the BAM file: Jurkat_K79_00p_Rep1_SRR1536557.sorted.bam
Done
[bam_sort_core] merging from 0 files and 8 in-memory blocks...
[bam_sort_core] merging from 0 files and 8 in-memory blocks...
[bam_sort_core] merging from 0 files and 8 in-memory blocks..

$ cat Jurkat_K79_00p_Rep1.report.txt

Sample n_unmapped      n_qcFail        n_duplicate     n_secondary     n_low_mapq      n_both  n_sample        n_exogenous
Jurkat_K79_00p_Rep1_SRR1536557.sorted.bam      719553  0       0       0       5355549 0       20741684        5207395
</pre></div>
</div>
<p>Mapping statistics and the derived <strong>scaling factors (SF)</strong> are summerized in the table below (the number of Drosophila reads in each sample was normalized to 1 million).</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 32%" />
<col style="width: 16%" />
<col style="width: 16%" />
<col style="width: 13%" />
<col style="width: 12%" />
<col style="width: 12%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Sample</p></td>
<td><p>Unmapped reads</p></td>
<td><p>Low_mapq_reads</p></td>
<td><p>Human_reads</p></td>
<td><p>Fly_reads</p></td>
<td><p>SF</p></td>
</tr>
<tr class="row-even"><td><p>Jurkat_K79_00p_Rep1_SRR1536557</p></td>
<td><p>719,553</p></td>
<td><p>5,355,549</p></td>
<td><p>20,741,684</p></td>
<td><p>5,207,395</p></td>
<td><p>0.1920346</p></td>
</tr>
<tr class="row-odd"><td><p>Jurkat_K79_25p_Rep1_SRR1536558</p></td>
<td><p>2,177,315</p></td>
<td><p>5,367,333</p></td>
<td><p>19,282,608</p></td>
<td><p>6,035,961</p></td>
<td><p>0.1656737</p></td>
</tr>
<tr class="row-even"><td><p>Jurkat_K79_50p_Rep1_SRR1536559</p></td>
<td><p>2,919,492</p></td>
<td><p>5,233,017</p></td>
<td><p>17,235,944</p></td>
<td><p>6,931,602</p></td>
<td><p>0.14426679</p></td>
</tr>
<tr class="row-odd"><td><p>Jurkat_K79_75p_Rep1_SRR1536560</p></td>
<td><p>1,627,494</p></td>
<td><p>4,300,454</p></td>
<td><p>11,731,507</p></td>
<td><p>8,291,870</p></td>
<td><p>0.12060006</p></td>
</tr>
<tr class="row-even"><td><p>Jurkat_K79_100p_Rep1_SRR1536561</p></td>
<td><p>4,526,721</p></td>
<td><p>5,109,981</p></td>
<td><p>7,893,377</p></td>
<td><p>14,372,323</p></td>
<td><p>0.06957817</p></td>
</tr>
<tr class="row-odd"><td><p>Jurkat_WCE_00p_Rep1.SRR1584489</p></td>
<td><p>2,344,210</p></td>
<td><p>3,914,206</p></td>
<td><p>11,126,768</p></td>
<td><p>844,395</p></td>
<td><p>1.18427987</p></td>
</tr>
<tr class="row-even"><td><p>Jurkat_WCE_25p_Rep1.SRR1584490</p></td>
<td><p>9,982,131</p></td>
<td><p>1,056,402</p></td>
<td><p>2,771,236</p></td>
<td><p>164,052</p></td>
<td><p>6.09562822</p></td>
</tr>
<tr class="row-odd"><td><p>Jurkat_WCE_50p_Rep1.SRR1584491</p></td>
<td><p>4,398,297</p></td>
<td><p>1,750,494</p></td>
<td><p>4,431,034</p></td>
<td><p>267,211</p></td>
<td><p>3.74236091</p></td>
</tr>
<tr class="row-even"><td><p>Jurkat_WCE_75p_Rep1.SRR1584492</p></td>
<td><p>7,317,599</p></td>
<td><p>1,620,854</p></td>
<td><p>4,029,096</p></td>
<td><p>179,641</p></td>
<td><p>5.56665795</p></td>
</tr>
<tr class="row-odd"><td><p>Jurkat_WCE_100p_Rep1.SRR1584493</p></td>
<td><p>7,609,117</p></td>
<td><p>2,336,349</p></td>
<td><p>5,197,374</p></td>
<td><p>344,665</p></td>
<td><p>2.901368</p></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You can also use K-mer based, alignment-free methods to calculate scaling factor. For example, you can use <a class="reference external" href="https://github.com/data61/gossamer/blob/master/docs/xenome.md">xenome</a> to separate human and fly reads from your FASTQ files.</p>
</div>
</div>
<div class="section" id="run-spiker">
<h1><span class="section-number">5. </span>Run Spiker<a class="headerlink" href="#run-spiker" title="Permalink to this headline">¶</a></h1>
<p><strong>Spiker</strong> is specifically designed to analyze ChIP-seq data with spike-in control. The input files can be single-end or
paired-end <a class="reference external" href="https://en.wikipedia.org/wiki/FASTQ_format#">FASTQ</a> files or pre-aligned <a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/bam.html">BAM</a> files. It supports both narrow peak (such as H3K27ac) and broad peak analysis (such as H3K79me2 in this tutorial).</p>
<p>In below, we used “Jurkat_K79_00p_Rep1_SRR1536557_human.sorted.bam” (ChIP) and “Jurkat_WCE_00p_Rep1.SRR1584489_human.sorted.bam” (control) as inputs to Spiker:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ spiker.py --broad --spikeIn --csf 1.18 --tsf 0.19 --bw -t Jurkat_K79_00p_Rep1_SRR1536557_human.sorted.bam -c Jurkat_WCE_00p_Rep1.SRR1584489_human.sorted.bam  -o Jurkat_K79_00p_Rep1_SRR1536557

2021-02-18 12:56:20 [INFO]  Running ChIP-seq workflow ...
2021-02-18 12:56:20 [INFO]  Input BAM files. Skip Step_1 (reads mapping).
2021-02-18 12:56:20 [INFO]  Step_2: Extract information from BAM files ...
2021-02-18 12:56:20 [INFO]  Step_2.1: Check sequencing layout from BAM file ...
2021-02-18 12:56:22 [INFO]     The layout of Jurkat_K79_00p_Rep1_SRR1536557_human.sorted.bam is SE
2021-02-18 12:56:23 [INFO]     The layout of Jurkat_WCE_00p_Rep1.SRR1584489_human.sorted.bam is SE
2021-02-18 12:56:23 [INFO]  Step_2.2: Extract information from ChIP BAM files ...
2021-02-18 12:56:25 [INFO]     The total mapped reads in ChIP sample is: 20741684
2021-02-18 12:56:25 [INFO]  Step_2.3: Extract information from control BAM files ...
2021-02-18 12:56:26 [INFO]     The total mapped reads in control sample is: 11126768
2021-02-18 12:56:26 [INFO]  Step_3: Deduplication ...
2021-02-18 12:56:26 [INFO]  Step_3.1: Remove duplidate reads from ChIP BAM file ...
...
</pre></div>
</div>
<p>Output files</p>
<dl class="glossary simple">
<dt id="term-.gappedPeak"><a href="#id3"><span class="problematic" id="id4">*</span></a>.gappedPeak</dt><dd><p>peak file (in <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format14">gappedPeak</a>) called by Spiker.</p>
</dd>
<dt id="term-.control.pileup.max.bdg"><a href="#id5"><span class="problematic" id="id6">*</span></a>.control.pileup.max.bdg</dt><dd><p><a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/bedgraph.html">bedGraph</a> file of <strong>control pileup</strong></p>
</dd>
<dt id="term-.control.pileup.max.bigWig"><a href="#id7"><span class="problematic" id="id8">*</span></a>.control.pileup.max.bigWig</dt><dd><p><a class="reference external" href="https://genome.ucsc.edu/goldenpath/help/bigWig.html">bigWig</a> file of <strong>control pileup</strong></p>
</dd>
<dt id="term-.treat.pileup.bdg"><a href="#id9"><span class="problematic" id="id10">*</span></a>.treat.pileup.bdg</dt><dd><p><a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/bedgraph.html">bedGraph</a> file of <strong>ChIP sample pileup</strong> (raw)</p>
</dd>
<dt id="term-.treat.pileup.SpikeIn_scaled.bdg"><a href="#id12"><span class="problematic" id="id13">*</span></a>.treat.pileup.SpikeIn_scaled.bdg</dt><dd><p><a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/bedgraph.html">bedGraph</a> file of <strong>ChIP sample pileup</strong> (Scaled to spike-in control)</p>
</dd>
<dt id="term-.treat.pileup.SpikeIn_scaled.bigWig"><a href="#id15"><span class="problematic" id="id16">*</span></a>.treat.pileup.SpikeIn_scaled.bigWig</dt><dd><p><a class="reference external" href="https://genome.ucsc.edu/goldenpath/help/bigWig.html">bigWig</a> file of <strong>ChIP sample pileup</strong> (Scaled to spike-in control)</p>
</dd>
</dl>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="bowtie2_index.html" class="btn btn-neutral float-right" title="1. Bowtie2 index files" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="usage.html" class="btn btn-neutral float-left" title="1. Spiker" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2021, Liguo Wang
      <span class="lastupdated">
        Last updated on Feb 12, 2021.
      </span>

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
    <!-- Theme Analytics -->
    <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-XXXXXXX-1', 'auto');
    ga('send', 'pageview');
    </script>

    
   

</body>
</html>